fingerprintdb_cuda.cu

```diff
@@ -143,43 +151,65 @@ void FingerprintDB::search (const Fingerprint& query,
     // Use Tanimoto to score similarity of all compounds to query fingerprint
     thrust::transform(d_results_indices.begin(), d_results_indices.end(),
             d_results_scores.begin(),
-            TanimotoFunctor(d_ref_fp, folded_fp_intsize, m_priv->d_data));
+            TanimotoFunctor(d_ref_fp, folded_fp_intsize, m_priv->d_data,
+                similarity_cutoff));
+
+    auto indices_end = d_results_indices.end();
+    auto scores_end = d_results_scores.end();
+    if(similarity_cutoff > 0) {
+        indices_end = thrust::remove_if(d_results_indices.begin(),
+                d_results_indices.end(), d_results_scores.begin(),
+                thrust::logical_not<bool>());
+        scores_end = thrust::remove(d_results_scores.begin(),
+                d_results_scores.end(), 0);
+    }
+    unsigned int indices_size = std::distance(d_results_indices.begin(),
+            indices_end);
 
     // Sort scores & indices vectors descending on score
-    thrust::sort_by_key(d_results_scores.begin(), d_results_scores.end(),
+    thrust::sort_by_key(d_results_scores.begin(), scores_end,
             d_results_indices.begin(), thrust::greater<float>());
+
+    results_to_consider = std::min(indices_size, return_count*m_fold_factor);
+
+    indices.assign(d_results_indices.begin(), 
+            d_results_indices.begin()+results_to_consider);
+
     } catch(thrust::system_error e) {
         cerr << "Error!  " << e.what() << endl;
     }
 
+
     if(m_fold_factor == 1) { // If we don't fold, we can take exact GPU results
         // Push top return_count results to CPU results vectors to be returned
-        for(unsigned int i=0;i<return_count*m_fold_factor;i++) {
+        for(unsigned int i=0;i<results_to_consider;i++) {
             results_smiles.push_back(m_smiles[d_results_indices[i]]);
             results_ids.push_back(m_ids[d_results_indices[i]]);
         }
         results_scores.assign(d_results_scores.begin(),
-                d_results_scores.begin()+return_count*m_fold_factor);
+                d_results_scores.begin()+results_to_consider);
     } else { // If we folded, we need to recalculate scores with full fingerprints
-        vector<int> indices(return_count*m_fold_factor);
-        results_scores.resize(return_count*m_fold_factor);
-        for(unsigned int i=0;i<return_count*m_fold_factor;i++) {
-            indices[i] = d_results_indices[i];
+        results_scores.resize(results_to_consider);
+        for(unsigned int i=0;i<results_to_consider;i++) {
             results_scores[i] = tanimoto_similarity_cpu(query,
                     getFingerprint(indices[i]));
-            cerr << results_scores[i] << " vs " << d_results_scores[i] << endl;
+            // Uncomment below to debug pre vs post folding scores
+            // cerr << results_scores[i] << " vs " << d_results_scores[i] << endl;
         }
         top_results_bubble_sort(indices, results_scores, return_count);
```

Comment:
Yeah, I changed it around so that it would sort in the direction we wanted, and avoided a lot of iterator mess around getting results from the end of the vector.  The main point of that bubble sort (and re-writing sort, which I'm loathe to do) is to get a constant sort time based on the amount of top results we care about.  *that said*, I wrote/used this when we were always returning 10 results.  Now that this engine might be used for 1000 results, bubble sort is going to be wayyyy too slow.  I'm glad you pointed this out :)